### PR TITLE
[DO NOT MERGE] TIP-689: Range attribute type

### PR DESCRIPTION
**Pull request by damien-carcel**
Working branch for testing the creation of a new attribute type. This is very basic stuff and is not to be merged.

### Current status

**What is working:**
- It is possible to create an attribute of type "range" programmatically, but not from the UI
- We can update/save a product programmatically with the product updater
- We can load it, then raw values are correctly converted as product values
- Versioning

**What is not:**
- PQB
- Localization
- PEF
- Product datagrid
- Completeness
- Mass edit
- Variant group
- Product export
- Product import

### CODE REVIEW

File: src/Acme/Bundle/RangeBundle/Factory/RangeFactory.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+use Acme\Bundle\RangeBundle\Model\Range;
+
+/**
+ * Creates a range object.
+ *
+ * @author    Damien Carcel (damien.carcel@akeneo.com)
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class RangeFactory
```
Comment (jjanvier) [inline src/Acme/Bundle/RangeBundle/Factory/RangeFactory.php:14]: You don't create a range here but a product value ;)

File: src/Acme/Bundle/RangeBundle/Model/Range.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    /** @var float */
+    protected $toData;
+
+    /**
+     * @param float $fromData
+     * @param float $toData
+     */
+    public function __construct($fromData, $toData)
+    {
+        $this->fromData = $fromData;
```
Comment (jjanvier) [inline src/Acme/Bundle/RangeBundle/Model/Range.php:24]: Maybe you could check the types of the data
And that from < to

File: src/Acme/Bundle/RangeBundle/Model/RangeProductValue.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+use Pim\Component\Catalog\Model\ProductValueInterface;
+
+/**
+ * A product value that can only contains range data.
+ *
+ * @author    Damien Carcel (damien.carcel@akeneo.com)
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class RangeProductValue extends AbstractProductValue implements ProductValueInterface
```
Comment (jjanvier) [inline src/Acme/Bundle/RangeBundle/Model/RangeProductValue.php:15]: You don't have to extend the product value normally now
That's one of the most significant enhancement we bring with the new model

Comment (jjanvier) [reply src/Acme/Bundle/RangeBundle/Model/RangeProductValue.php:15]: Ok my bad
I think you still need it at the moment

File: src/Acme/Bundle/RangeBundle/Normalizer/Flat/RangeNormalizer.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+use Acme\Bundle\RangeBundle\Model\Range;
+use Pim\Bundle\VersioningBundle\Normalizer\Flat\AbstractProductValueDataNormalizer;
+
+/**
+ * @author    Damien Carcel (damien.carcel@akeneo.com)
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class RangeNormalizer extends AbstractProductValueDataNormalizer
+{
```
Comment (jjanvier) [inline src/Acme/Bundle/RangeBundle/Normalizer/Flat/RangeNormalizer.php:14]: What is it used for?

File: src/Acme/Bundle/RangeBundle/Normalizer/Storage/Product/RangeNormalizer.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+use Acme\Bundle\RangeBundle\Model\Range;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+/**
+ * @author    Damien Carcel (damien.carcel@akeneo.com)
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class RangeNormalizer implements NormalizerInterface
```
Comment (jjanvier) [inline src/Acme/Bundle/RangeBundle/Normalizer/Storage/Product/RangeNormalizer.php:13]: Too bad we have to do this one
Maybe we could provide a basic normalizer for storage based on the standard one

**Review by jjanvier — COMMENTED** [review]
Good work so far 👏
Let see how it evolve with the coming features and evolutions on the product model

File: src/Acme/Bundle/RangeBundle/Model/Range.php (hunk 2 above)
Comment (damien-carcel) [reply src/Acme/Bundle/RangeBundle/Model/Range.php:24]: I could, but imo, not here but in the RangeFactory.

File: src/Acme/Bundle/RangeBundle/Factory/RangeFactory.php (hunk 1 above)
Comment (damien-carcel) [reply src/Acme/Bundle/RangeBundle/Factory/RangeFactory.php:14]: No, I create a Range object, like for price and metrics. Then I use the RangeProductValueFactory that call this factory.

File: src/Acme/Bundle/RangeBundle/Model/RangeProductValue.php (hunk 3 above)
Comment (damien-carcel) [reply src/Acme/Bundle/RangeBundle/Model/RangeProductValue.php:15]: I could have made my own product value, but I would had problem with the interface, so it was simpler this way.

File: src/Acme/Bundle/RangeBundle/Normalizer/Flat/RangeNormalizer.php (hunk 4 above)
Comment (damien-carcel) [reply src/Acme/Bundle/RangeBundle/Normalizer/Flat/RangeNormalizer.php:14]: Versionning

File: src/Acme/Bundle/RangeBundle/Normalizer/Storage/Product/RangeNormalizer.php (hunk 5 above)
Comment (damien-carcel) [reply src/Acme/Bundle/RangeBundle/Normalizer/Storage/Product/RangeNormalizer.php:13]: You mean one standard normalizer that support both "standard" and "storage"? Didn't even think of it, I admit, I just kept it the more similar to our existing attribute types.

**Pull request closed without merging**